internal/impl/tigerbeetle/input_tigerbeetle.go

```diff
@@ -0,0 +1,701 @@
+func (input *tigerbeetleInput) produce(ctx context.Context, client tb.Client, timestampLast uint64) error {
+	// Asynchronously closes the client,
+	// forcing any inflight request to finish in case of hard stop.
+	go func() {
+		select {
+		case <-input.stopSignaller.HasStoppedChan(): // Graceful shutdown.
+		case <-input.stopSignaller.HardStopChan(): // Hard stop.
+		}
+		client.Close()
+	}()
+
+	idleTimer := time.NewTimer(0)
+	_ = idleTimer.Stop()
+
+	for {
+		err := input.checkRateLimit(ctx)
+		if err != nil {
+			return err
+		}
+
+		input.logger.Debugf("producer: get_change_events: timestamp_min=%d limit=%d",
+			timestampLast+1,
+			input.config.eventCountMax,
+		)
+
+		results, err := client.GetChangeEvents(tb_types.ChangeEventsFilter{
+			TimestampMin: timestampLast + 1,
+			TimestampMax: 0,
+			Limit:        input.config.eventCountMax,
+		})
+		if err != nil {
+			return err
+		}
+
+		input.logger.Debugf("producer: get_change_events: %d results", len(results))
+
+		// No events returned from the query,
+		// waiting for the timeout to resume the producer.
+		if len(results) == 0 {
+			// NB: We could go idle if `len(results) < eventCountMax`, since the client
+			// likely won’t return new results if queried again immediately.
+			// However, we wait for the *consumer* to begin flushing the current results
+			// before issuing a new query, avoiding unnecessary idle time for workloads
+			// with high frequency but low volume per batch.
+			rescheduled := idleTimer.Reset(input.config.idleInterval)
+			if rescheduled {
+				return errors.New("assertion failed: idle timer was already running")
+			}
+
+			input.logger.Debugf("producer: idle: %d ms", input.config.idleInterval.Milliseconds())
+
+			select {
+			case <-idleTimer.C:
+				continue
+			case <-ctx.Done():
+				_ = idleTimer.Stop()
+				return ctx.Err()
+			}
+		}
+
+		// Waits until the consumer flushes the results or the job is stopped.
+		select {
+		case input.producerChan <- results:
+			timestampLast = results[len(results)-1].Timestamp
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// Flushes the events into the pipeline.
+func (input *tigerbeetleInput) consume(ctx context.Context) error {
+	// We must keep events ordered,
+	// the next batch can only be flushed when the current one has been acknowledged.
+	batch := make([]*service.Message, 0, input.config.eventCountMax)
+	ackChan := make(chan struct{}, 1)
+	for {
+		select {
+		case results := <-input.producerChan:
+			if len(results) == 0 {
+				return errors.New("assertion failed: unexpected empty results")
+			} else if len(results) > int(input.config.eventCountMax) {
+				return errors.New("assertion failed: too many results")
+			} else if len(batch) != 0 {
+				return errors.New("assertion failed: pending messages to flush")
+			}
+
+			for _, result := range results {
+				bytes, err := jsonSerialize(result)
+				if err != nil {
+					input.logger.Errorf("Unable to serialize as JSON: %w", err)
```

Comment:
This should be %s as we're logging. That said, given the code is returning an error below which is eventually logged as part of the errorgroup, perhaps returning a wrapped error with the same error message would suffice to reduce double logging?